### PR TITLE
#2303 - Total eligible amount in NOA screen

### PR DESCRIPTION
**Pull request by guru-aot**
Total eligible amount is not showing for NOA (part-time and full-time)

image

image

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 1]
```diff
@@ -152,6 +155,25 @@ export class AssessmentControllerService {
...
 
+  /**
+   * Calculate the sum of 'valueAmount' for each DisbursementValue within an array of DisbursementSchedules.
+   * @param disbursementSchedules disbursement schedule details.
+   * @returns The total sum of 'valueAmount' across all DisbursementSchedules.
+   */
+  private sumDisbursementValueAmounts(
+    disbursementSchedules: DisbursementSchedule[],
+  ): number {
+    return disbursementSchedules.reduce((totalValueAmount, schedule) => {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:166]: The logic could be simplified like this.

```ts
 return disbursementSchedules
    .flatMap((disbursement) => disbursement.disbursementValues)
    .map((disbursementValue) => disbursementValue.valueAmount)
    .reduce((accumulator, currentValue) => accumulator + currentValue, 0);
```

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 2]
```diff
@@ -152,6 +155,25 @@ export class AssessmentControllerService {
     return disbursementDetails;
   }
 
+  /**
+   * Calculate the sum of 'valueAmount' for each DisbursementValue within an array of DisbursementSchedules.
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:159]: Can avoid technical terms(like array) and object names DisbursementSchedules in the comment.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/models/assessment.dto.ts [hunk 3]
```diff
@@ -65,6 +65,7 @@ export class AssessmentNOAAPIOutDTO {
   offeringIntensity: OfferingIntensity;
   offeringStudyStartDate: string;
   offeringStudyEndDate: string;
+  eligibleAmount: number;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/models/assessment.dto.ts:68]: Something like `totalEligibleAward` could be more aligned with the purpose.

Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/models/assessment.dto.ts:68]: Not a blocker. I see that eligibleAmount is used in NOA form.

File: sources/packages/web/src/services/http/dto/Assessment.dto.ts [hunk 4]
```diff
@@ -58,6 +58,7 @@ export interface AssessmentNOAAPIOutDTO {
   offeringStudyEndDate: string;
   msfaaNumber: string;
   disbursement: unknown;
+  eligibleAmount: number;
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/Assessment.dto.ts:61]: I see that following fields are used in NOA form.io but not saved by the workflow. 

Are these fields not relevant to the part time application?

 
image

image

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 5]
```diff
@@ -65,21 +66,158 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+    const [firstDisbursementSchedule, secondDisbursementSchedule] =
+      assessment.disbursementSchedules;
 
-    const awards = {};
-    firstDisbursementSchedule.disbursementValues.forEach((disbursement) => {
-      awards[`disbursement1${disbursement.valueCode.toLowerCase()}`] =
-        disbursement.valueAmount;
-    });
+    const firstDisbursementScheduleAwards = {};
+    firstDisbursementSchedule.disbursementValues.forEach(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:87]: minor. can use a common  method to create disbursement schedule awards.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 1 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:166]: I agree, it can be done in many different ways, but i believe for nested arrays scenarios and summing a value, i believe reduce is generally more efficient and appropriate. Also i thought, of creating a private method to have the return in the getAssessmentNOA method cleaner. 
Do you still believe this needs to be changed?

File: sources/packages/web/src/services/http/dto/Assessment.dto.ts (hunk 4 above)
Comment (guru-aot) [reply sources/packages/web/src/services/http/dto/Assessment.dto.ts:61]: Yes if you are talking about the form.io not saving the assessmentData by the PartTime application, yes there are fields that are used only for FullTime application.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 6]
```diff
@@ -65,21 +66,158 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+          assessment.offering.studyEndDate,
+        ),
+        offeringStudyStartDate: getDateOnlyFormat(
+          assessment.offering.studyStartDate,
+        ),
+        programName: assessment.offering.educationProgram.name,
+      });
+  });
+
+  it("Should get the student noa details for an eligible partime application when an eligible public institution user tries to access it.", async () => {
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:178]: minor typo `part time`, one `t` missing

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 7]
```diff
@@ -65,21 +66,158 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+    const student = await saveFakeStudent(db.dataSource);
+
+    const currentMSFAA = createFakeMSFAANumber(
+      { student },
+      {
+        msfaaState: MSFAAStates.Signed,
+      },
+    );
+    await db.msfaaNumber.save(currentMSFAA);
+    const application = await saveFakeApplicationDisbursements(
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:191]: I believe the default intensity is full time, since this test is for part time, shouldn't we make the application `part time`?

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 8]
```diff
@@ -65,21 +66,158 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+      db.dataSource,
+      {
+        institution: collegeF,
+        institutionLocation: collegeFLocation,
+        student,
+        msfaaNumber: currentMSFAA,
+      },
+      {
+        offeringIntensity: OfferingIntensity.fullTime,
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:79]: since this test is for full-time, maybe we can add it to the test description to avoid confusion

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 9]
```diff
@@ -34,3 +34,18 @@ export function createFakeDisbursementSchedule(relations?: {
   schedule.disbursementScheduleStatus = DisbursementScheduleStatus.Pending;
   return schedule;
 }
+
+export function createDisbursementScheduleAwards(
+  disbursementSchedule,
+  disbursementNumber,
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:40]: pls add the types

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 10]
```diff
@@ -34,3 +34,18 @@ export function createFakeDisbursementSchedule(relations?: {
...
+) {
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:41]: pls add a return type

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 11]
```diff
@@ -34,3 +34,18 @@ export function createFakeDisbursementSchedule(relations?: {
   schedule.disbursementScheduleStatus = DisbursementScheduleStatus.Pending;
   return schedule;
 }
+
+export function createDisbursementScheduleAwards(
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:38]: pls add method comments

**Review by ann-aot — COMMENTED** [review]
Added  some comments

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 1 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:166]: I don't have any hard feelings but I like Dheepak's suggestion cleaner.

**Review by andrepestana-aot — APPROVED** [review]
Good job! I'm approving as I don't see any major issues.

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 12]
```diff
@@ -34,3 +34,31 @@ export function createFakeDisbursementSchedule(relations?: {
   schedule.disbursementScheduleStatus = DisbursementScheduleStatus.Pending;
   return schedule;
 }
+
+/**
+ * Disbursement awards.
+ */
+interface DisbursementAwards {
+  [disbursementValueCode: string]: number;
+}
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:43]: Not sure, if we are adding the interface in the same file in `api/`, if we are then pls move it to the top or move it to the respective file

**Review by ann-aot — COMMENTED** [review]
LGTM @guru-aot , just a minor comment

**Review by ann-aot — APPROVED** [review]
NIce work @guru-aot . 👍 Thanks for doing the changes

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:166]: @guru-aot I still feel that it would make the code cleaner and easy to grasp.

File: sources/packages/web/src/services/http/dto/Assessment.dto.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/Assessment.dto.ts:61]: My question is, does part time noa require Living Allowance, Other allowable cost and Total Assessed Cost ? or they are disregarded part time NOA. 

I am doubtful that they may be needed for PT noa as well. But I may be wrong. 

But if they are needed, the value is never going to come for PT application unless I am missing something.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:166]: We have the same approach suggested by @dheepak-aot in other places and it was exactly the same that I mentioned over our call about the ticket @guru-aot maybe it was not clear during that time.
Please apply the suggestion since it seems cleaner.

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 13]
```diff
@@ -34,3 +41,24 @@ export function createFakeDisbursementSchedule(relations?: {
...
+ */
+export function createDisbursementScheduleAwards(
+  disbursementSchedule: DisbursementSchedule,
+  disbursementNumber: number,
+): DisbursementAwards {
+  const disbursementScheduleAwards: DisbursementAwards = {};
+
+  disbursementSchedule.disbursementValues.forEach((disbursementValue) => {
+    disbursementScheduleAwards[
+      `disbursement${disbursementNumber}${disbursementValue.valueCode.toLowerCase()}`
```
Comment (sh16011993) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:59]: Is there a possibility of having a repeated `valueCode` creating a key conflict (duplicate key)?

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 14]
```diff
@@ -34,3 +41,24 @@ export function createFakeDisbursementSchedule(relations?: {
...
   return schedule;
 }
+
+/**
+ * Creates disbursement schedule awards based on the given schedule and disbursement number.
+ * @param disbursementSchedule Disbursement schedule containing disbursement values.
+ * @param disbursementNumber Number associated with the disbursement schedule (e.g., 1 for the first, 2 for the second).
+ * @returns Disbursement awards with keys like 'disbursement1code' and values as amounts.
+ */
+export function createDisbursementScheduleAwards(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:51]: Is this generating the API payload for the disbursement awards, right?
This is an API-specific method for a specific controller. Maybe it can be restricted to that controller's methods or even a local method if it fits the current scenario.

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts [hunk 15]
```diff
@@ -34,3 +41,24 @@ export function createFakeDisbursementSchedule(relations?: {
...
+  disbursementSchedule: DisbursementSchedule,
+  disbursementNumber: number,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:53]: The parameter can be declared in a way that would restrict the value to the spectated ones, for instance.
```ts
disbursementNumber: 1 | 2,
```

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts (hunk 14 above)
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:51]: Should it be named something explicit to NOA, for instance, `createDisbursementScheduleAwardsForNOA` or `createNOADisbursementScheduleAwards`?

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 16]
```diff
@@ -65,21 +67,147 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+            secondDisbursementSchedule.msfaaNumber?.dateSigned,
+          disbursement2MSFAAId: secondDisbursementSchedule.msfaaNumber?.id,
+          disbursement2MSFAANumber: MASKED_MSFAA_NUMBER,
+          disbursement2Status:
+            secondDisbursementSchedule.disbursementScheduleStatus,
+          disbursement2TuitionRemittance:
+            secondDisbursementSchedule.tuitionRemittanceRequestedAmount,
+          ...secondDisbursementScheduleAwards,
+        },
+        eligibleAmount: assessment.disbursementSchedules.reduce(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:146]: I have two points here.

First, when doing this type of E2E test we can also consider the option to control the values saved to the DB (the factories allow it) and then do the expected check as, for instance, 
```ts
eligibleAmount: 984321 // Sum of all controlled values manually calculated based on the DB values manually assigned.
```

Second, if we are not controlling the values as suggested on the "First" point (which I would prefer) I would consider creating a local method similar to the `sumDisbursementValueAmounts`.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes. 👍

Wanted to know the outcome of the comment https://github.com/bcgov/SIMS/pull/2365/files#r1346112384 from @JasonCTang @michesmith

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 17]
```diff
@@ -65,21 +67,147 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
-      application.currentAssessment.disbursementSchedules;
+    const [firstDisbursementSchedule, secondDisbursementSchedule] =
+      assessment.disbursementSchedules;
 
-    const awards = {};
-    firstDisbursementSchedule.disbursementValues.forEach((disbursement) => {
-      awards[`disbursement1${disbursement.valueCode.toLowerCase()}`] =
-        disbursement.valueAmount;
-    });
+    const firstDisbursementScheduleAwards = createDisbursementScheduleAwards(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:87]: Regarding the method `createDisbursementScheduleAwards`, IMO, it can also fall into the "First" point mentioned on https://github.com/bcgov/SIMS/pull/2365/files#r1346571749.
In general, for E2E test assertions, if possible to control the data, I would recommend to have more hard-coded `expect` verifications.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 18]
```diff
@@ -152,6 +155,25 @@ export class AssessmentControllerService {
...
+    disbursementSchedules: DisbursementSchedule[],
+  ): number {
+    return disbursementSchedules
+      .flatMap(
+        (disbursementSchedule) => disbursementSchedule.disbursementValues,
+      )
+      .reduce(
+        (accumulator, disbursementValue) =>
+          accumulator + disbursementValue.valueAmount,
+        0,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:173]: IMO: Specifying `0` as the initial value explicitly to the `reduce` is not required here.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts [hunk 19]
```diff
@@ -112,8 +240,36 @@ describe("AssessmentInstitutionsController(e2e)-getAssessmentNOA", () => {
...
+            secondDisbursementSchedule.tuitionRemittanceRequestedAmount,
+          ...secondDisbursementScheduleAwards,
         },
+        eligibleAmount: assessment.disbursementSchedules.reduce(
+          (totalValueAmount, schedule) => {
+            return (
+              totalValueAmount +
+              (schedule.disbursementValues || []).reduce(
+                (sum, disbursementValue) => sum + disbursementValue.valueAmount,
+                0,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts:267]: https://github.com/bcgov/SIMS/pull/2365/files#r1346581248

**Review by sh16011993 — COMMENTED** [review]
Nice work @guru-aot 👍 Added an info and a minor suggestion comment.

File: sources/packages/web/src/services/http/dto/Assessment.dto.ts (hunk 4 above)
Comment (guru-aot) [reply sources/packages/web/src/services/http/dto/Assessment.dto.ts:61]: I have confirmed with @HRAGANBC regarding this and there will be a future ticket that needs displaying of several other fields in the NOA screen. Will connect the ticket with out ticket for continuation.

File: sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts (hunk 13 above)
Comment (guru-aot) [reply sources/packages/backend/libs/test-utils/src/factories/disbursement-schedule.ts:59]: No there will not be any

File: sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts (hunk 19 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/_tests_/e2e/assessment.institutions.controller.getAssessmentNOA.e2e-spec.ts]: Removed it

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 18 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.controller.service.ts:173]: Creating an accumulator in the start, in case if the array is null, then the value returned will be 0.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍
